homeworks/olena.ambitska_elena-ambitska/GamesProject/scripts/login.js

```diff
@@ -0,0 +1,107 @@
+        required: true,
+    },
+    checkbox: {
+        rule: (checkbox) => checkbox,
+        message: 'Checkbox must be checked',
+        required: true,
+    },
+};
+
+const ERROR_STATE = Object.fromEntries(Object.keys(VALIDATE_RULES).map((key) => [key, false]));
+
+function displayErrorMessage(field, message) {
+    const errorContainer = document.createElement('div');
+    errorContainer.classList.add('error-message');
+    errorContainer.setAttribute('data-field', field);
+    errorContainer.innerText = message;
+
+    const inputField = document.getElementById(field);
+    const formGroup = inputField.closest('.form-group');
+    formGroup.appendChild(errorContainer);
+}
+
+function validateField(field) {
+    const value = document.getElementById(field).value.trim();
+    const rule = VALIDATE_RULES[field];
+    const isChecked = document.getElementById(field).checked;
+
+    const isFieldEmpty = rule.required && !value;
+    const isCheckboxInvalid = field === 'checkbox' && !rule.rule(isChecked);
+    const isValueInvalid = rule.rule && !rule.rule(value);
+
+    // eslint-disable-next-line object-curly-newline
+    return { value, rule, isChecked, isFieldEmpty, isCheckboxInvalid, isValueInvalid };
+}
+
+function clearErrorMessages(field) {
+    if (field) {
+        const errorMessages = document.querySelectorAll(`.error-message[data-field="${field}"]`);
+        errorMessages.forEach((errorMessage) => errorMessage.remove());
+        ERROR_STATE[field] = false;
+    } else {
+        const errorMessages = document.querySelectorAll('.error-message');
+        errorMessages.forEach((errorMessage) => errorMessage.remove());
+        Object.keys(VALIDATE_RULES).forEach((key) => {
+            ERROR_STATE[key] = false;
+        });
+    }
+}
+
+function validateInputField(field) {
+    clearErrorMessages(field);
+
+    // eslint-disable-next-line object-curly-newline
+    const { rule, isFieldEmpty, isCheckboxInvalid, isValueInvalid } = validateField(field);
+
+    if ((isFieldEmpty || isCheckboxInvalid || isValueInvalid) && !ERROR_STATE[field]) {
+        displayErrorMessage(field, isFieldEmpty ? 'This field is required' : rule.message);
+        ERROR_STATE[field] = true;
+    } else if (!isFieldEmpty && !isCheckboxInvalid && !isValueInvalid) {
+        clearErrorMessages(field);
+        ERROR_STATE[field] = false;
+    }
+
+    document.getElementById('submit-button').disabled = Object.values(ERROR_STATE).some((error) => error);
+}
+
+function validateFormHandler(event) {
+    event.preventDefault();
+
+    clearErrorMessages();
+
+    let isValid = true;
+
+    Object.keys(VALIDATE_RULES).forEach((field) => {
+        const { isFieldEmpty, isCheckboxInvalid, isValueInvalid } = validateField(field);
+
+        validateInputField(field);
+
+        if (isFieldEmpty || isCheckboxInvalid || isValueInvalid) {
+            isValid = false;
+        }
+    });
+
+    document.getElementById('submit-button').disabled = !isValid;
```

Comment:
You have the global `ERROR_STATE` properly filled at this moment. So I propose to get `isValid` for it. To do so, you can skip lines 84 and 88. And you would not call `validateField` twice. Here I've logged invocation of all your handlers and just hit submit on empty form
![image](https://github.com/MastersAcademy/fe-course-basic-2023/assets/98031768/69faa9f8-c9ef-4a44-b56f-890c766f49be)

Reply:
@vladyslav-yermolin-moc Good catch! fixed!

![image](https://github.com/MastersAcademy/fe-course-basic-2023/assets/74259586/0d7b8778-829b-46ae-9d91-fd1a1f6b0571)